### PR TITLE
Replace kernel resource strings with real .cu source files

### PR DESCRIPTION
**Pull request by tlemo**
The new setup (as implemented in this PR) has 3 key parts:

1. Build system changes to copy the files from under `.../jit/codegen/cuda/runtime` to the build output location (and also installing them during a full setup)
2. At runtime, detect the location of the library binary (ex. `torch_cuda.so` or `torch_cuda.dll`). See `torchLibPath()` for details.
3. Pass `--include-path=` to NVRTC to point to the right location (relative to the library location from step 2)

After this, we can simply `#include <nvfuser_runtime/foo.cu>` from the generated code.

### CODE REVIEW

**Comment by csarofeen** [comment]
@malfet We're trying a bit of a tweak on our build. There are files that we would like included in the install that we can use like headers to JIT our code. We've tested it on Linux and Windows but would really appreciate some feedback. We isolated this out from all our other work to make it easier for you to check it out and give us feedback.

**Review by malfet — COMMENTED** [review]
This makes the feature more fragile, less resilient to relocations.
Perhaps a better approach would be to generate `kernel_resource_strings.h` at compile time and then instead of materializing those headers on disk pass it as `headers` and `includeNames` parameters of `nvrtcCreateProgram`, as described in
https://docs.nvidia.com/cuda/nvrtc/index.html#group__compilation_1g9ae65f68911d1cf0adda2af4ad8cb458

File: torch/csrc/jit/codegen/cuda/runtime/tensor.cu [hunk 1]
```diff
@@ -0,0 +1,27 @@
+
+typedef unsigned char uint8_t;
+typedef signed char int8_t;
+typedef short int int16_t;
+typedef long long int int64_t;
+
+template <typename T, int N>
```
[the remark this answers is not in the document]
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/runtime/tensor.cu:7]: Good question, although the content of the .cu files is extracted from the old resource strings and not in scope for significant changes in this PR (unless we spot severe issues, of course)

File: torch/csrc/jit/codegen/cuda/runtime/tensor.cu [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+struct Tensor {
+  __device__ T& operator[](int64_t ind) {
+    return data[ind];
+  };
+
+  T* data;
+  int64_t size[N];
```
[the remark this answers is not in the document]
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/runtime/tensor.cu:14]: ditto

File: torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    unsigned int lo1 = mulhilo32(kPhiloxSB, ctr.z, &hi1);
+    uint4 ret = {hi1 ^ ctr.y ^ key.x, lo1, hi0 ^ ctr.w ^ key.y, lo0};
+    return ret;
+  }
+
+ private:
+  static const unsigned long kPhilox10A = 0x9E3779B9;
+  static const unsigned long kPhilox10B = 0xBB67AE85;
+  static const unsigned long kPhiloxSA = 0xD2511F53;
+  static const unsigned long kPhiloxSB = 0xCD9E8D57;
```
[the remark this answers is not in the document]
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu:94]: Good suggestion. Again, the content of the .cu files is lifted from the original resource strings.

The `constexpr` change is safe enough so I'll incorporate it, but I'll leave the base type alone for now (NB `int` is not guaranteed to be 32bit)

**Comment by tlemo** [comment]
> This makes the feature more fragile, less resilient to relocations.
> Perhaps a better approach would be to generate `kernel_resource_strings.h` at compile time and then instead of materializing those headers on disk pass it as `headers` and `includeNames` parameters of `nvrtcCreateProgram`, as described in
> https://docs.nvidia.com/cuda/nvrtc/index.html#group__compilation_1g9ae65f68911d1cf0adda2af4ad8cb458

Thanks for the feedback! We considered that approach as well, but the ability to patch the `.cu` files without a pre-build step & recompilation is very attractive (both for local development and in the field).

At this point the main goal is to move away from the hard-coded strings. Do you see any concrete problems that would prevent us from using the solution in this PR? Otherwise it would help to merge it in sooner rather than later (We can reconsider the pre-build preprocessing approach if needed. The nice thing is that switching from resource files to pre-processing would not require us to touch the .cu source files)

**Comment by malfet** [comment]
> At this point the main goal is to move away from the hard-coded strings. Do you see any concrete problems that would prevent us from using the solution in this PR? Otherwise it would help to merge it in sooner rather than later (We can reconsider the pre-build preprocessing approach if needed. The nice thing is that switching from resource files to pre-processing would not require us to touch the .cu source files)
This approach would only work if libtorch is a dynamic dependency. But that would not be the case, if application is linked statically with libtorch using PyTorch C++ API.

Correct me if I'm wrong, but the chage you are proposing are aimed at making development cycle faster for nvFuser developers rather than for end user of the feature? If that is the case, then perhaps right solution would be to check for certain environment variable (say `TORCH_CUDA_CODEGEN_RUNTIME_PATH`) and only in that case switch from embedded headers to an on-disk location.  This would avoid any potential deployment regressions but at the same time would allow feature developers do fast iterations without recompiling libtorch_cuda every time.

**Comment by tlemo** [comment]
> This approach would only work if libtorch is a dynamic dependency. But that would not be the case, if application is linked statically with libtorch using PyTorch C++ API.

This is a good point, although the solution in this PR works with static linking as well (the runtime path makes no assumptions about the name of the module, see `torchLibPath` in executor_utils.cpp). The only requirement is that the `nvfuser_runtime` subdirectory is located next to the binary.

Just curious, how common is static linking used with libtorch? Is there any test case for it in the PyTorch code base?

> Correct me if I'm wrong, but the chage you are proposing are aimed at making development cycle faster for nvFuser developers rather than for end user of the feature? If that is the case, then perhaps right solution would be to check for certain environment variable (say `TORCH_CUDA_CODEGEN_RUNTIME_PATH`) and only in that case switch from embedded headers to an on-disk location. This would avoid any potential deployment regressions but at the same time would allow feature developers do fast iterations without recompiling libtorch_cuda every time.

Sorry if the explanation was not clear: the main goal is to switch from hardcoding CUDA code in string literals to using "first class" .cu files (since the former is quickly becoming unmanageable). 

We have evaluated two options: A. treat .cu files as resource files (this PR) and B. preprocess .cu files to produce something that can be linked in during build (the idea you mentioned). The ability to patch the code w/o recompiling is just one of the reasons we decided to go with A. Also, patching where recompilation is not an easy option ("in-the-field") is really the more attractive part.

Moreover, we anticipate other uses for external resource files (ex. pre-compiled CUDA libraries to be used with `cuLinkAddFile`), so having a mechanism in place is useful.

**Pull request closed without merging**